[HIP] Add extra exception handling for non-ROCM builds

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #83009
* #80072

I got the following error on OSX, which doesn't have HIP.
As this file is supposed to compile with non-HIP builds,
I added this error to the errors to ignore.

```
Traceback (most recent call last):
  File "test/test_profiler.py", line 31, in <module>
    from torch.profiler._pattern_matcher import (Pattern, NamePattern,
  File "/Users/jclow/pytorch3/torch/profiler/_pattern_matcher.py", line 9, in <module>
    import torch.utils.benchmark as benchmark
  File "/Users/jclow/pytorch3/torch/utils/benchmark/__init__.py", line 2, in <module>
    from torch.utils.benchmark.utils.timer import *  # noqa: F403
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/timer.py", line 8, in <module>
    from torch.utils.benchmark.utils import common, cpp_jit
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/cpp_jit.py", line 13, in <module>
    from torch.utils import cpp_extension
  File "/Users/jclow/pytorch3/torch/utils/cpp_extension.py", line 19, in <module>
    from .hipify import hipify_python
  File "/Users/jclow/pytorch3/torch/utils/hipify/hipify_python.py", line 34, in <module>
    from .cuda_to_hip_mappings import CUDA_TO_HIP_MAPPINGS
  File "/Users/jclow/pytorch3/torch/utils/hipify/cuda_to_hip_mappings.py", line 34, in <module>
    rocm_path = subprocess.check_output(["hipconfig", "--rocmpath"]).decode("utf-8")
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: 'hipconfig'
```

Differential Revision: [D38766067](https://our.internmc.facebook.com/intern/diff/D38766067)

Gamrix committed:
[HIP] Add extra exception handling for non-ROCM builds

I got the following error on OSX, which doesn't have HIP.
As this file is supposed to compile with non-HIP builds,
I added this error to the errors to ignore.

```
Traceback (most recent call last):
  File "test/test_profiler.py", line 31, in <module>
    from torch.profiler._pattern_matcher import (Pattern, NamePattern,
  File "/Users/jclow/pytorch3/torch/profiler/_pattern_matcher.py", line 9, in <module>
    import torch.utils.benchmark as benchmark
  File "/Users/jclow/pytorch3/torch/utils/benchmark/__init__.py", line 2, in <module>
    from torch.utils.benchmark.utils.timer import *  # noqa: F403
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/timer.py", line 8, in <module>
    from torch.utils.benchmark.utils import common, cpp_jit
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/cpp_jit.py", line 13, in <module>
    from torch.utils import cpp_extension
  File "/Users/jclow/pytorch3/torch/utils/cpp_extension.py", line 19, in <module>
    from .hipify import hipify_python
  File "/Users/jclow/pytorch3/torch/utils/hipify/hipify_python.py", line 34, in <module>
    from .cuda_to_hip_mappings import CUDA_TO_HIP_MAPPINGS
  File "/Users/jclow/pytorch3/torch/utils/hipify/cuda_to_hip_mappings.py", line 34, in <module>
    rocm_path = subprocess.check_output(["hipconfig", "--rocmpath"]).decode("utf-8")
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: 'hipconfig'
```

[ghstack-poisoned]

Gamrix committed:
Update on "[HIP] Add extra exception handling for non-ROCM builds"


I got the following error on OSX, which doesn't have HIP.
As this file is supposed to compile with non-HIP builds,
I added this error to the errors to ignore.

```
Traceback (most recent call last):
  File "test/test_profiler.py", line 31, in <module>
    from torch.profiler._pattern_matcher import (Pattern, NamePattern,
  File "/Users/jclow/pytorch3/torch/profiler/_pattern_matcher.py", line 9, in <module>
    import torch.utils.benchmark as benchmark
  File "/Users/jclow/pytorch3/torch/utils/benchmark/__init__.py", line 2, in <module>
    from torch.utils.benchmark.utils.timer import *  # noqa: F403
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/timer.py", line 8, in <module>
    from torch.utils.benchmark.utils import common, cpp_jit
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/cpp_jit.py", line 13, in <module>
    from torch.utils import cpp_extension
  File "/Users/jclow/pytorch3/torch/utils/cpp_extension.py", line 19, in <module>
    from .hipify import hipify_python
  File "/Users/jclow/pytorch3/torch/utils/hipify/hipify_python.py", line 34, in <module>
    from .cuda_to_hip_mappings import CUDA_TO_HIP_MAPPINGS
  File "/Users/jclow/pytorch3/torch/utils/hipify/cuda_to_hip_mappings.py", line 34, in <module>
    rocm_path = subprocess.check_output(["hipconfig", "--rocmpath"]).decode("utf-8")
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: 'hipconfig'
```

[ghstack-poisoned]

malfet reviewed (approved):
Looks good to me, but why we even running this one on MacOS?

Gamrix commented:
The thing that I find most suspect is the topline  import of 
`from .hipify import hipify_python` 
in 
`torch/utils/cpp_extension.py`

I don't see any need for that module to be imported by default. 

The change was introduced in https://github.com/pytorch/pytorch/pull/35897 . @ezyang , as the reviewer of that diff, do you know of any reason why we should have HIP import be a top level import for this file?

ezyang commented:
If you want to hide it behind actual use when we notice we're compiling a hipify that's fine. But imports are supposed to be fine at the top level unless they result in an import cycle.

malfet commented:
Since this subprocess call is only triggered when one imports `torch.utils.cpp_extension` this is probably fine, though still a bit ugly

Gamrix commented:
Seems like the best solution should be to not do the call to check if HIP is available in the module scope, and instead call it when used and cache the result.

Gamrix commented:
@pytorchmergebot merge

Gamrix committed:
Update on "[HIP] Add extra exception handling for non-ROCM builds"


I got the following error on OSX, which doesn't have HIP.
As this file is supposed to compile with non-HIP builds,
I added this error to the errors to ignore.

```
Traceback (most recent call last):
  File "test/test_profiler.py", line 31, in <module>
    from torch.profiler._pattern_matcher import (Pattern, NamePattern,
  File "/Users/jclow/pytorch3/torch/profiler/_pattern_matcher.py", line 9, in <module>
    import torch.utils.benchmark as benchmark
  File "/Users/jclow/pytorch3/torch/utils/benchmark/__init__.py", line 2, in <module>
    from torch.utils.benchmark.utils.timer import *  # noqa: F403
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/timer.py", line 8, in <module>
    from torch.utils.benchmark.utils import common, cpp_jit
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/cpp_jit.py", line 13, in <module>
    from torch.utils import cpp_extension
  File "/Users/jclow/pytorch3/torch/utils/cpp_extension.py", line 19, in <module>
    from .hipify import hipify_python
  File "/Users/jclow/pytorch3/torch/utils/hipify/hipify_python.py", line 34, in <module>
    from .cuda_to_hip_mappings import CUDA_TO_HIP_MAPPINGS
  File "/Users/jclow/pytorch3/torch/utils/hipify/cuda_to_hip_mappings.py", line 34, in <module>
    rocm_path = subprocess.check_output(["hipconfig", "--rocmpath"]).decode("utf-8")
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: 'hipconfig'
```

[ghstack-poisoned]

Gamrix commented:
[Windows tests are failing due to unrelated changes https://github.com/pytorch/pytorch/issues/83433 ]

Gamrix committed:
Update on "[HIP] Add extra exception handling for non-ROCM builds"


I got the following error on OSX, which doesn't have HIP.
As this file is supposed to compile with non-HIP builds,
I added this error to the errors to ignore.

```
Traceback (most recent call last):
  File "test/test_profiler.py", line 31, in <module>
    from torch.profiler._pattern_matcher import (Pattern, NamePattern,
  File "/Users/jclow/pytorch3/torch/profiler/_pattern_matcher.py", line 9, in <module>
    import torch.utils.benchmark as benchmark
  File "/Users/jclow/pytorch3/torch/utils/benchmark/__init__.py", line 2, in <module>
    from torch.utils.benchmark.utils.timer import *  # noqa: F403
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/timer.py", line 8, in <module>
    from torch.utils.benchmark.utils import common, cpp_jit
  File "/Users/jclow/pytorch3/torch/utils/benchmark/utils/cpp_jit.py", line 13, in <module>
    from torch.utils import cpp_extension
  File "/Users/jclow/pytorch3/torch/utils/cpp_extension.py", line 19, in <module>
    from .hipify import hipify_python
  File "/Users/jclow/pytorch3/torch/utils/hipify/hipify_python.py", line 34, in <module>
    from .cuda_to_hip_mappings import CUDA_TO_HIP_MAPPINGS
  File "/Users/jclow/pytorch3/torch/utils/hipify/cuda_to_hip_mappings.py", line 34, in <module>
    rocm_path = subprocess.check_output(["hipconfig", "--rocmpath"]).decode("utf-8")
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/jclow/opt/anaconda3/envs/pytorch3/lib/python3.8/subprocess.py", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: 'hipconfig'
```

[ghstack-poisoned]

Gamrix commented:
@pytorchmergebot merge

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38766067).

Gamrix commented:
@pytorchmergebot merge

Gamrix commented:
@pytorchbot merge

kit1980 commented on torch/utils/hipify/cuda_to_hip_mappings.py:
```diff
@@ -34,7 +34,7 @@
 except subprocess.CalledProcessError:
     print(f"Warning: hipconfig --rocmpath failed, assuming {rocm_path}")
-except FileNotFoundError:
+except (FileNotFoundError, PermissionError):
```
@Gamrix How this can change anything? `PermissionError` was already handled with `pass` in the next lines.

Gamrix replied:
Ugh, I think multiple people were fixing this at the same time. That line below was not there when I originally hit the error :/
